ext/opcache/jit/zend_jit_arm64.dasc

```diff
@@ -0,0 +1,6050 @@
+		    opline->result_type != IS_UNUSED) {
+			|	brk #0  // TODO: test
+			|	ZVAL_COPY_VALUE res_addr, res_use_info, op1_def_addr, MAY_BE_LONG, ZREG_TMP1, ZREG_TMP2, ZREG_TMP3, ZREG_TMP4, ZREG_FPTMP1
+		}
+		|.cold_code
+		|1:
+		|	brk #0  // TODO: test
+		|	b >3
+		|.code
+	} else {
+		if ((opline->opcode == ZEND_PRE_INC || opline->opcode == ZEND_PRE_DEC) &&
+		    opline->result_type != IS_UNUSED) {
+			|	ZVAL_COPY_VALUE res_addr, res_use_info, op1_def_addr, MAY_BE_LONG, ZREG_TMP1, ZREG_TMP2, ZREG_TMP3, ZREG_TMP4, ZREG_FPTMP1
+		}
+	}
+	if (op1_info & ((MAY_BE_ANY|MAY_BE_UNDEF)-MAY_BE_LONG)) {
+		|.cold_code
+		|2:
+		|	brk #0  // TODO: test
+		|	b >3
+		|.code
+	}
+	|3:
+	if (!zend_jit_store_var_if_necessary_ex(Dst, opline->op1.var, op1_def_addr, op1_def_info, op1_addr, op1_info)) {
+		return 0;
+	}
+	if (opline->result_type != IS_UNUSED) {
+		if (!zend_jit_store_var_if_necessary(Dst, opline->result.var, res_addr, res_info)) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static int zend_jit_opline_uses_reg(const zend_op  *opline, int8_t reg)
+{
+	if ((opline+1)->opcode == ZEND_OP_DATA
+	 && ((opline+1)->op1_type & (IS_VAR|IS_TMP_VAR|IS_CV))
+	 && JIT_G(current_frame)->stack[EX_VAR_TO_NUM((opline+1)->op1.var)].reg == reg) {
+		return 1;
+	}
+	return
+		((opline->result_type & (IS_VAR|IS_TMP_VAR|IS_CV)) &&
+			JIT_G(current_frame)->stack[EX_VAR_TO_NUM(opline->result.var)].reg == reg) ||
+		((opline->op1_type & (IS_VAR|IS_TMP_VAR|IS_CV)) &&
+			JIT_G(current_frame)->stack[EX_VAR_TO_NUM(opline->op1.var)].reg == reg) ||
+		((opline->op2_type & (IS_VAR|IS_TMP_VAR|IS_CV)) &&
+			JIT_G(current_frame)->stack[EX_VAR_TO_NUM(opline->op2.var)].reg == reg);
+}
+
+static int zend_jit_math_long_long(dasm_State    **Dst,
+                                   const zend_op  *opline,
+                                   zend_uchar      opcode,
+                                   zend_jit_addr   op1_addr,
+                                   zend_jit_addr   op2_addr,
+                                   zend_jit_addr   res_addr,
+                                   uint32_t        res_info,
+                                   uint32_t        res_use_info,
+                                   int             may_overflow)
+{
+	bool same_ops = zend_jit_same_addr(op1_addr, op2_addr);
+	zend_reg result_reg;
+
+	// x86 defines a 'tmp_reg' to handle integer overflow case.
+	// In AArch64, we directly use our reserved TMP1.
+	// zend_reg tmp_reg = ZREG_X0;
+
+	if (Z_MODE(res_addr) == IS_REG && (res_info & MAY_BE_LONG)) {
+		if (may_overflow && (res_info & MAY_BE_GUARD)
+		 && JIT_G(current_frame)
+		 && zend_jit_opline_uses_reg(opline, Z_REG(res_addr))) {
+			result_reg = ZREG_TMP3;	// to store the result temporarily. Use TMP3
+		} else {
+			result_reg = Z_REG(res_addr);
+		}
+	} else if (Z_MODE(op1_addr) == IS_REG && Z_LAST_USE(op1_addr)) {
+		|	brk #0	// TODO: test
+		result_reg = Z_REG(op1_addr);
+	} else if (Z_REG(res_addr) != ZREG_X0) {
```

Comment:
IS_REG is handled above (on line 2577). Here we handle IS_MEM_ZVAL with [R0 + 0], which can come only from ASSIGN_DIM_OP. You probably should use one of your reserved temporary registers instead of R0 and another reserved temporary register instead of FCARG1.

Reply:
First, sorry for the confusing line numbers I used in my previous comment.
Thanks for your explanation. But I'm afraid I didn't get you.
1) As you mentioned, IS_REG is handled on line 2577, hence the condition `(res_info & MAY_BE_LONG)` is useless, right?
2) Regarding ASSIGN_DIM_OP, I found there are a number of special handling for this opcode, and all of them are located in math-related handlers, i.e. zend_jit_math_*().
As you mentioned, [R0 + 0] can come only from ASSIGN_DIM_OP? But I wonder where can I find such convention? As you suggest, I should use one of my reserved temp register there.
3) since we use our reserved temp register instead of FCARG1, do you think the def scratch register at line 5876 is not needed?

Reply:
> First, sorry for the confusing line numbers I used in my previous comment.
> Thanks for your explanation. But I'm afraid I didn't get you.
> 
>     1. As you mentioned, IS_REG is handled on line 2577, hence the condition `(res_info & MAY_BE_LONG)` is useless, right?

No. It's possible situation when we ADD two integers and get DOUBLE result in XMM register. This is possible only with tracing JIT and guards.

>     2. Regarding ASSIGN_DIM_OP, I found there are a number of special handling for this opcode, and all of them are located in math-related handlers, i.e. zend_jit_math_*().

right.

>        As you mentioned, [R0 + 0] can come only from ASSIGN_DIM_OP? But I wonder where can I find such convention? As you suggest, I should use one of my reserved temp register there.

This is just an unique way how we pass address of temporary zval from ASSIGN_DIM_OP. There are a number of /* ASSIGN_DIM_OP */ comments in zend_jit_math_*(). This trick is mainly used to improve register allocation on x86 with just few available registers.
 
>     3. since we use our reserved temp register instead of FCARG1, do you think the def scratch register at line 5876 is not needed?

yes. not needed.

Reply:
Oh I see. Thanks a lot for your explanation. Will update.